app/models/concerns/one_time_password_checkable.rb

```diff
@@ -1,31 +1,35 @@
 module OneTimePasswordCheckable
   extend ActiveSupport::Concern
 
+  CATEGORIES = %w[
+    claim_email
+    claim_mobile
+    reminder_email
+  ].freeze
+
   included do
     attribute :one_time_password, :string, limit: 6
+    attribute :one_time_password_category, :string
     attribute :sent_one_time_password_at, :datetime
-    validate :otp_validate, on: [:"email-verification"]
-    before_save :set_sent_one_time_password_at, if: :persisted?
+    validate :otp_validate, on: [:"email-verification", :"mobile-verification"]
     before_save :normalise_one_time_password, if: :one_time_password_changed?
   end
 
   private
 
-  def set_sent_one_time_password_at
-    self.sent_one_time_password_at = sent_one_time_password_at
-  end
-
   def normalise_one_time_password
     self.one_time_password = one_time_password.gsub(/\D/, "")
   end
 
   def otp_validate
-    return write_attribute(:email_verified, true) if otp.valid?
+    return unless CATEGORIES.include?(one_time_password_category)
+    return write_attribute(:email_verified, true) if otp.valid? && %w[claim_email reminder_email].include?(one_time_password_category)
+    return write_attribute(:mobile_verified, true) if otp.valid? && one_time_password_category == "claim_mobile"
 
     errors.add(:one_time_password, otp.warning)
   end
 
   def otp
-    @otp ||= OneTimePassword::Validator.new(one_time_password, set_sent_one_time_password_at)
+    @otp ||= OneTimePassword::Validator.new(one_time_password, sent_one_time_password_at, nil, one_time_password_category)
```

Comment:
rather than pass in `nil` to `OneTimePassword::Validator` I'd say get rid of the `ISSUER` setting in `Validator` and `Generator` it doesn't look like it is needed. I copied this over from the original code.I thought it was used in validation but if you can just pass in `nil` and it still works, then I guess not!